src/main/scala/zio/telemetry/Telemetry.scala

```diff
@@ -0,0 +1,213 @@
+package zio.telemetry
+
+import io.opentracing.Span
+import io.opentracing.SpanContext
+import io.opentracing.Tracer
+import io.opentracing.propagation.Format
+import io.opentracing.tag.Tag
+import java.util.concurrent.TimeUnit
+import scala.jdk.CollectionConverters._
+import zio.Exit
+import zio.FiberRef
+import zio.IO
+import zio.Task
+import zio.UIO
+import zio.ZIO
+import zio.ZManaged
+import zio.clock.Clock
+
+trait Telemetry extends Serializable {
+  def telemetry: Telemetry.Service
+}
+
+object Telemetry {
+
+  trait Service {
+    def spanRef: FiberRef[Span]
+    def tracer: Tracer
+  }
+
+  def managedService(
+      tracer: Tracer,
+      rootOpName: String = "ROOT"
+  ): ZManaged[Clock, Nothing, Telemetry.Service] =
+    ZManaged.make(
+      for {
+        span <- UIO(tracer.buildSpan(rootOpName).start())
+        ref <- FiberRef.make(span)
+        tracer_ = tracer
+      } yield new Telemetry.Service {
+        override def spanRef: FiberRef[Span] = ref
+        override def tracer: Tracer = tracer_
+      }
+    )(_.spanRef.get.flatMap(span => UIO(span.finish)))
+
+  type COT = Clock with Telemetry
+
+  def underlying[R, R1 <: R with Telemetry, E, A](
+      f: Tracer => ZIO[R, E, A]
+  ): ZIO[R1, E, A] =
+    getTracer.flatMap(f)
+
+  def spanFrom[R, R1 <: R with COT, E, A, C <: Object](
+      format: Format[C],
+      carrier: C,
+      zio: ZIO[R, E, A],
+      opName: String,
+      tagError: Boolean = true,
+      logError: Boolean = true
+  ): ZIO[R1, E, A] =
+    getTracer.flatMap { tracer =>
+      Task(tracer.extract(format, carrier))
+        .fold(_ => None, Option.apply)
+        .flatMap {
+          case None => zio
```

Comment:
I'd fold this.

Reply:
With fold we need explicit type parameters. In that case, I prefer pattern matching.